Dockerfile

```diff
@@ -239,26 +239,64 @@ RUN xx-apk add musl-dev gcc && xx-go --wrap
 
 FROM containerd-base AS containerd
 ARG CONTAINERD_VERSION
-RUN --mount=from=containerd-src,src=/usr/src/containerd,readwrite --mount=target=/root/.cache,type=cache \
-  git fetch origin \
-  && git checkout -q "$CONTAINERD_VERSION" \
-  && make bin/containerd \
-  && make bin/containerd-shim-runc-v2 \
-  && make bin/ctr \
-  && mv bin /out
+RUN --mount=from=containerd-src,src=/usr/src/containerd,rw \
+    --mount=target=/root/.cache,type=cache <<EOT
+  set -e
+  git fetch origin
+  git checkout -q "$CONTAINERD_VERSION"
+  mkdir /out
+  if [ "$(xx-info os)" = "windows" ]; then
+    CGO_ENABLED=0 make STATIC=1 binaries
+    mv bin/containerd.exe bin/containerd-shim* bin/ctr* /out
+  else
+    make bin/containerd
+    make bin/containerd-shim-runc-v2
+    make bin/ctr
+    mv bin/containerd bin/containerd-shim* bin/ctr* /out
+  fi
+EOT
+
+FROM scratch AS containerd-binaries
+COPY --link --from=containerd /out /
 
 # containerd v1.6 for integration tests
 FROM containerd-base as containerd-alt-16
 ARG CONTAINERD_ALT_VERSION_16
-RUN --mount=from=containerd-src,src=/usr/src/containerd,readwrite --mount=target=/root/.cache,type=cache \
-  git fetch origin \
-  && git checkout -q "$CONTAINERD_ALT_VERSION_16" \
-  && make bin/containerd \
-  && make bin/containerd-shim-runc-v2 \
-  && mv bin /out
+RUN --mount=from=containerd-src,src=/usr/src/containerd,rw \
+    --mount=target=/root/.cache,type=cache <<EOT
+  set -e
+  git fetch origin
+  git checkout -q "$CONTAINERD_ALT_VERSION_16"
+  mkdir /out
+  if [ "$(xx-info os)" = "windows" ]; then
+    CGO_ENABLED=0 make STATIC=1 binaries
+    mv bin/containerd.exe bin/containerd-shim* /out
+  else
+    make bin/containerd
+    make bin/containerd-shim-runc-v2
+    mv bin/containerd bin/containerd-shim* /out
+  fi
+EOT
 
+FROM scratch AS containerd-alt-16-binaries
+COPY --link --from=containerd-alt-16 /out/containerd* /
+
+FROM gobuild-base AS registry
 ARG REGISTRY_VERSION
-FROM registry:$REGISTRY_VERSION AS registry
+RUN git clone https://github.com/distribution/distribution.git /go/src/github.com/docker/distribution
+WORKDIR /go/src/github.com/docker/distribution
+ARG TARGETPLATFORM
+RUN --mount=target=/root/.cache,type=cache \
+  git checkout -q "$REGISTRY_VERSION" && \
+  export GOPATH="/go/src/github.com/docker/distribution/Godeps/_workspace:$GOPATH" && \
+  GO111MODULE=off CGO_ENABLED=0 xx-go build -o /registry ./cmd/registry && \
+  xx-verify --static /registry
+
+FROM scratch AS registry-binaries-linux
+COPY --link --from=registry /registry /
+FROM scratch AS registry-binaries-windows
+COPY --link --from=registry /registry /registry.exe
```

Comment:
Looks like this could be done with a wildcard or empty `/out` destination directory